### ISSUE TITLE
[Phase 3] Constrained Modeling and Calibration

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/credit-scoring-reject-inference#4]
## Summary

Implement constrained logistic regression with monotonicity constraints, probability calibration (Platt scaling and beta calibration), and scorecard conversion. This ensures model coefficients align with domain knowledge and probabilities are well-calibrated.

**Part of:** #1 (Meta Issue)
**Depends on:** #3 (Phase 2: Reject Inference)

## System Context

```
src/modeling/
├── __init__.py
├── constrained_logistic.py   # Monotonic logistic regression (THIS PHASE)
├── calibration.py            # Platt/beta calibration (THIS PHASE)
└── scorecard.py              # Points-based scorecard (THIS PHASE)

tests/
└── test_modeling.py          # Unit tests (THIS PHASE)
```

### Why Monotonicity Constraints?

In credit scoring, certain relationships must be monotonic for regulatory approval:
- Higher income → Lower default probability (negative coefficient)
- Higher debt ratio → Higher default probability (positive coefficient)
- More delinquencies → Higher default probability (positive coefficient)

Unconstrained models might learn spurious relationships that violate these expectations.

### Why Calibration?

Raw model probabilities often don't match true frequencies:
- Model predicts 30% default, but actual default rate is 40%
- Critical for regulatory capital calculations and pricing

## Objectives

- [ ] Implement constrained logistic regression with sign constraints
- [ ] Use SciPy SLSQP optimizer for constrained optimization
- [ ] Implement Platt scaling calibration
- [ ] Implement beta calibration for asymmetric miscalibration
- [ ] Create scorecard converter (log-odds to points)

### Constrained Logistic Regression

📄 **File:** `src/modeling/constrained_logistic.py`

```python
import numpy as np
from scipy.optimize import minimize
from scipy.special import expit  # Sigmoid
from typing import Dict, Optional, List


class MonotonicLogisticRegression:
    """Logistic regression with monotonicity constraints."""

    def __init__(self, constraints: Optional[Dict[str, str]] = None,
                 C: float = 1.0, max_iter: int = 1000):
        """
        Args:
            constraints: Dict mapping feature name to constraint type.
                        'positive': coefficient must be >= 0
                        'negative': coefficient must be <= 0
                        e.g., {'income': 'negative', 'debt_ratio': 'positive'}
            C: Regularization strength (L2)
            max_iter: Maximum optimizer iterations
        """
        self.constraints = constraints or {}
        self.C = C
        self.max_iter = max_iter
        self.coef_: Optional[np.ndarray] = None
        self.intercept_: float = 0.0
        self.feature_names_: List[str] = []

    def fit(self, X: np.ndarray, y: np.ndarray,
            feature_names: Optional[List[str]] = None) -> 'MonotonicLogisticRegression':
        """Fit with constrained optimization."""
        n_samples, n_features = X.shape

        if feature_names is None:
            feature_names = [f'x{i}' for i in range(n_features)]
        self.feature_names_ = feature_names

        # Initial parameters (coefficients + intercept)
        x0 = np.zeros(n_features + 1)

        # Build constraints
        scipy_constraints = self._build_constraints(n_features)

        # Optimize
        result = minimize(
            fun=lambda params: self._objective(params, X, y),
            x0=x0,
            method='SLSQP',
            constraints=scipy_constraints,
            options={'maxiter': self.max_iter, 'disp': False}
        )

        if not result.success:
            print(f"Warning: Optimization did not converge: {result.message}")

        self.coef_ = result.x[:-1]
        self.intercept_ = result.x[-1]
        return self

    def _objective(self, params: np.ndarray, X: np.ndarray,
                   y: np.ndarray) -> float:
        """Negative log-likelihood with L2 regularization."""
        coef = params[:-1]
        intercept = params[-1]

        z = X @ coef + intercept
        proba = expit(z)

        # Log-likelihood (with numerical stability)
        ll = np.sum(
            y * np.log(proba + 1e-10) +
            (1 - y) * np.log(1 - proba + 1e-10)
        )

        # L2 regularization
        reg = 0.5 * self.C * np.sum(coef ** 2)

        return -(ll - reg)

    def _build_constraints(self, n_features: int) -> list:
        """Build SciPy constraint dictionaries."""
        constraints = []

        for i, name in enumerate(self.feature_names_):
            if name in self.constraints:
                if self.constraints[name] == 'positive':
                    # coef[i] >= 0
                    constraints.append({
                        'type': 'ineq',
                        'fun': lambda params, idx=i: params[idx]
                    })
                elif self.constraints[name] == 'negative':
                    # coef[i] <= 0 equivalent to -coef[i] >= 0
                    constraints.append({
                        'type': 'ineq',
                        'fun': lambda params, idx=i: -params[idx]
                    })

        return constraints

    def predict_proba(self, X: np.ndarray) -> np.ndarray:
        """Predict class probabilities."""
        z = X @ self.coef_ + self.intercept_
        proba_1 = expit(z)
        return np.column_stack([1 - proba_1, proba_1])

    def predict(self, X: np.ndarray, threshold: float = 0.5) -> np.ndarray:
        """Predict class labels."""
        return (self.predict_proba(X)[:, 1] >= threshold).astype(int)
```

### Probability Calibration

📄 **File:** `src/modeling/calibration.py`

```python
import numpy as np
from scipy.optimize import minimize
from scipy.special import expit


class PlattCalibrator:
    """
    Platt scaling for probability calibration.

    Fits sigmoid: P(y=1|s) = 1 / (1 + exp(A*s + B))
    where s is the raw model output (log-odds or probability).
    """

    def __init__(self):
        self.A_: float = 1.0
        self.B_: float = 0.0

    def fit(self, probas: np.ndarray, y_true: np.ndarray) -> 'PlattCalibrator':
        """
        Fit sigmoid calibration parameters.

        Args:
            probas: Raw probabilities from model
            y_true: True binary labels
        """
        # Convert to log-odds for fitting
        log_odds = np.log(probas / (1 - probas + 1e-10) + 1e-10)

        def objective(params):
            A, B = params
            calibrated = expit(-(A * log_odds + B))
            # Cross-entropy loss
            loss = -np.mean(
                y_true * np.log(calibrated + 1e-10) +
                (1 - y_true) * np.log(1 - calibrated + 1e-10)
            )
            return loss

        result = minimize(objective, x0=[1.0, 0.0], method='BFGS')
        self.A_, self.B_ = result.x
        return self

    def transform(self, probas: np.ndarray) -> np.ndarray:
        """Apply calibration."""
        log_odds = np.log(probas / (1 - probas + 1e-10) + 1e-10)
        return expit(-(self.A_ * log_odds + self.B_))


class BetaCalibrator:
    """
    Beta calibration for asymmetric miscalibration.

    Transforms probabilities using: calibrated = a + b * p^c
    where a, b, c are fitted parameters.
    """

    def __init__(self):
        self.a_: float = 0.0
        self.b_: float = 1.0
        self.c_: float = 1.0

    def fit(self, probas: np.ndarray, y_true: np.ndarray) -> 'BetaCalibrator':
        """Fit beta calibration parameters."""
        def objective(params):
            a, b, c = params
            # Ensure valid probability output
            c = max(c, 0.01)
            calibrated = np.clip(a + b * np.power(probas + 1e-10, c), 1e-10, 1 - 1e-10)
            loss = -np.mean(
                y_true * np.log(calibrated) +
                (1 - y_true) * np.log(1 - calibrated)
            )
            return loss

        result = minimize(
            objective,
            x0=[0.0, 1.0, 1.0],
            method='L-BFGS-B',
            bounds=[(-1, 1), (0, 5), (0.01, 5)]
        )
        self.a_, self.b_, self.c_ = result.x
        return self

    def transform(self, probas: np.ndarray) -> np.ndarray:
        """Apply beta calibration."""
        return np.clip(
            self.a_ + self.b_ * np.power(probas + 1e-10, self.c_),
            0, 1
        )
```

### Scorecard Converter

📄 **File:** `src/modeling/scorecard.py`

```python
import numpy as np
from typing import Dict


class ScorecardConverter:
    """
    Convert logistic regression to points-based scorecard.

    Score = Offset - Factor * log_odds
    where log_odds = intercept + sum(coef * WoE)
    """

    def __init__(self, base_score: int = 600, base_odds: float = 50,
                 pdo: int = 20):
        """
        Args:
            base_score: Score at base odds (e.g., 600)
            base_odds: Odds at base score (e.g., 50:1 non-default)
            pdo: Points to double odds
        """
        self.base_score = base_score
        self.base_odds = base_odds
        self.pdo = pdo
        self.factor = pdo / np.log(2)
        self.offset = base_score - self.factor * np.log(base_odds)

    def log_odds_to_score(self, log_odds: np.ndarray) -> np.ndarray:
        """Convert log-odds to credit score."""
        return self.offset - self.factor * log_odds

    def score_to_probability(self, score: np.ndarray) -> np.ndarray:
        """Convert score to default probability."""
        log_odds = (self.offset - score) / self.factor
        return 1 / (1 + np.exp(log_odds))

    def generate_scorecard(self, coef: np.ndarray, intercept: float,
                           feature_names: list,
                           woe_maps: Dict[str, Dict]) -> Dict:
        """
        Generate points for each category.

        Returns dict: {feature: {category: points}}
        """
        scorecard = {}

        # Distribute intercept points across features
        n_features = len(feature_names)
        intercept_contribution = intercept / n_features

        for i, (name, woe_map) in enumerate(woe_maps.items()):
            scorecard[name] = {}
            for category, woe in woe_map.items():
                # Points = -Factor * (coef * WoE + intercept/n)
                log_odds_contrib = coef[i] * woe + intercept_contribution
                points = -self.factor * log_odds_contrib
                scorecard[name][category] = round(points)

        return scorecard
```

## Definition of Done

- [ ] Constrained optimization respects sign constraints
- [ ] Calibration improves Brier score on validation set
- [ ] Reliability diagram shows improved calibration curve
- [ ] Scorecard produces scores in 300-850 range
- [ ] All tests pass with coverage > 85%

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/modeling/constrained_logistic.py` | 1-100 | Create | Monotonic logistic regression |
| `src/modeling/calibration.py` | 1-80 | Create | Platt and beta calibration |
| `src/modeling/scorecard.py` | 1-60 | Create | Points-based scorecard |
| `tests/test_modeling.py` | 1-100 | Create | Unit tests |

## Resources

- [SciPy SLSQP Documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html)
- [Platt Scaling Paper](https://www.cs.colorado.edu/~mozer/Teaching/syllabi/6622/papers/Platt1999.pdf)
- [Credit Scorecard Development](https://www.listendata.com/2015/09/credit-scorecard-development.html)
